Source/gendung.cpp

```diff
@@ -30,12 +30,7 @@ std::optional<OwnedCelSprite> pSpecialCels;
 std::unique_ptr<MegaTile[]> pMegaTiles;
 std::unique_ptr<uint16_t[]> pLevelPieces;
 std::unique_ptr<byte[]> pDungeonCels;
-std::array<uint8_t, MAXTILES + 1> block_lvid;
-std::array<bool, MAXTILES + 1> nBlockTable;
-std::array<bool, MAXTILES + 1> nSolidTable;
-std::array<bool, MAXTILES + 1> nTransTable;
-std::array<bool, MAXTILES + 1> nMissileTable;
-std::array<bool, MAXTILES + 1> nTrapTable;
+std::unique_ptr<TileProperties[]> SOLData;

```

Comment:
Is this moved to the heap in order to remove the `MAXTILES` limitation?

Reply:
My understanding of the various allocation types is still pretty weak. If i understand correctly fixed sized globals are pre allocated in a static block during boot up, but dynamically allocated memory would have to request general memory (heap?) from the OS at run time, making it the slowest.

I understand the stack as part of the statically allocated memory, reserved for variables used locally in functions. Meaning that the application can do it's own memory management with out context switching as long as it can fit within this size.

I actually thought loading the SOL data in a local variable would still result in a heap allocation, the only difference here being that it was de-allocated once the function returned. But you comment suggest to me that it would have been able to use the stack? If that is the case I guess there would be some logic to decide to do so only if it's small enough to fit in the stack?

I'm not sure what your question actually relates to, but I guess more of the lookup could have been per-calculated by the compiler if done against the static memory rather then the heap? If you think that would make a difference we can change it to be a fixed size, doing so would still end up saving us 10kb (down from 11kb if my math is correct).

Reply:
I actually have no idea where the MAXTILES limit comes from, unless .cel has an inherent limit to how many frames it can contain I think this is just an arbitrary number and we are free to pick our own as long as it is large enough for the largest tileset (town at 1376).

Reply:
`std::array<T, N>` is equivalent to a `T[N]`, so it's in the static memory.

> I actually thought loading the SOL data in a local variable would still result in a heap allocation

Previously the data was loaded into a static array, so there was no heap allocation going on there (other than the file I/O internals which may have some). There was no local variable there. If there were a local `std::array` and it was too large to fit on the stack, the program would simply crash (`std::array` is always stack-allocated).

If we don't plan to increase `MAXTILES`, it's better to change it from `std::array` to `std::unique_ptr`.

Reply:
Nevermind, it was indeed temporarily heap-allocated previously (I misread the code)